pkg/controller/dataimportcron-controller_test.go

```diff
@@ -0,0 +1,315 @@
+/*
+Copyright 2021 The CDI Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	imagev1 "github.com/openshift/api/image/v1"
+
+	v1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/tools/record"
+
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+)
+
+var (
+	cronLog         = logf.Log.WithName("data-import-cron-controller-test")
+	cronName        = "test-cron"
+	imageStreamName = "test-imagestream"
+)
+
+const (
+	testDigest    = "sha256:68b44fc891f3fae6703d4b74bcc9b5f24df8d23f12e642805d1420cbe7a4be70"
+	testDockerRef = "quay.io/kubevirt/blabla@" + testDigest
+)
+
+var _ = Describe("All DataImportCron Tests", func() {
+	var _ = Describe("DataImportCron controller reconcile loop", func() {
+		var (
+			reconciler *DataImportCronReconciler
+			cronKey    = types.NamespacedName{Name: cronName, Namespace: metav1.NamespaceDefault}
+			cronReq    = reconcile.Request{NamespacedName: cronKey}
+			cronJobKey = func(cron *cdiv1.DataImportCron) types.NamespacedName {
+				return types.NamespacedName{Name: GetCronJobName(cron), Namespace: reconciler.cdiNamespace}
+			}
+			dataSourceKey = func(cron *cdiv1.DataImportCron) types.NamespacedName {
+				return types.NamespacedName{Name: cron.Spec.ManagedDataSource, Namespace: metav1.NamespaceDefault}
+			}
+			dvKey = func(dvName string) types.NamespacedName {
+				return types.NamespacedName{Name: dvName, Namespace: metav1.NamespaceDefault}
+			}
+		)
+		AfterEach(func() {
+			if reconciler != nil {
+				close(reconciler.recorder.(*record.FakeRecorder).Events)
+				reconciler = nil
+			}
+		})
+
+		It("Should do nothing and return nil when no DataImportCron exists", func() {
+			reconciler = createDataImportCronReconciler()
+			_, err := reconciler.Reconcile(context.TODO(), cronReq)
+			Expect(err).ToNot(HaveOccurred())
+		})
+
+		DescribeTable("Should", func(setDeletionTimestamp bool) {
```

Comment:
I would just make this an "It" should create and delete cronjob if dataimport cron is created and delete. In the second loop you are basically running the first loop again, and then deleting. Its not a huge deal. The tables are mostly useful for if you are running the same function with various inputs and various expectations. Here you are creating a import cron and verify the job is created, and then deleting (why not just delete it instead of setting the timestamp?) and checking that the cron is gone.

Reply:
fixed